src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.DecryptTokenResult.cs

```diff
@@ -0,0 +1,151 @@
+        /// <returns>The decrypted, and if the 'zip' claim is set, decompressed string representation of the token.</returns>
+        internal static TokenDecryptingResult DecryptJwtToken(
+            JsonWebToken jsonWebToken,
+            ValidationParameters validationParameters,
+            JwtTokenDecryptionParameters decryptionParameters,
+            CallContext callContext)
+        {
+            if (validationParameters == null)
+                return new TokenDecryptingResult(
+                    jsonWebToken,
+                    ValidationFailureType.NullArgument,
+                    new ExceptionDetail(
+                        new MessageDetail(
+                            TokenLogMessages.IDX10000,
+                            nameof(validationParameters)),
+                        typeof(ArgumentNullException),
+                        new System.Diagnostics.StackFrame()));
+
+            if (decryptionParameters == null)
+                return new TokenDecryptingResult(
+                    jsonWebToken,
+                    ValidationFailureType.NullArgument,
+                    new ExceptionDetail(
+                        new MessageDetail(
+                            TokenLogMessages.IDX10000,
+                            nameof(decryptionParameters)),
+                        typeof(ArgumentNullException),
+                        new System.Diagnostics.StackFrame()));
+
+            bool decryptionSucceeded = false;
+            bool algorithmNotSupportedByCryptoProvider = false;
+            byte[] decryptedTokenBytes = null;
+
+            // keep track of exceptions thrown, keys that were tried
+            StringBuilder exceptionStrings = null;
+            StringBuilder keysAttempted = null;
+            string zipAlgorithm = null;
+            foreach (SecurityKey key in decryptionParameters.Keys)
+            {
+                var cryptoProviderFactory = validationParameters.CryptoProviderFactory ?? key.CryptoProviderFactory;
+                if (cryptoProviderFactory == null)
+                {
+                    if (LogHelper.IsEnabled(EventLogLevel.Warning))
+                        LogHelper.LogWarning(TokenLogMessages.IDX10607, key);
+
+                    continue;
+                }
+
+                try
+                {
+                    if (!cryptoProviderFactory.IsSupportedAlgorithm(jsonWebToken.Enc, key))
+                    {
+                        if (LogHelper.IsEnabled(EventLogLevel.Warning))
+                            LogHelper.LogWarning(TokenLogMessages.IDX10611, LogHelper.MarkAsNonPII(decryptionParameters.Enc), key);
+
+                        algorithmNotSupportedByCryptoProvider = true;
+                        continue;
+                    }
+
+                    AlgorithmValidationResult result = validationParameters.AlgorithmValidator(zipAlgorithm, key, jsonWebToken, validationParameters, callContext);
```

Comment:
Consider adding a List of AlgorithmValidationResult and attaching to the TokenDecryptionResult (renamed), then no need to have the exceptionStrings.

Reply:
Adding a work item to look at this after we have our `ValidateTokenAsync` implementation working.